Improve Container Types

Mozz3d opened the pull request.

Mozz3d committed:
Adjust file paths

Mozz3d committed:
Improve Container Types

Mozz3d committed:
Fix include path

Mozz3d committed:
Adjustments

wopss commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,498 @@
+        return Data()[aPos];
+    }
+
+    [[nodiscard]] constexpr Reference At(std::make_signed_t<SizeType> aPos)
```
```suggestion
    [[nodiscard]] constexpr Reference At(SizeType aPos)
```

wopss commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,498 @@
+        return Data()[static_cast<SizeType>(aPos)];
+    }
+
+    [[nodiscard]] constexpr ConstReference At(std::make_signed_t<SizeType> aPos) const
```
```suggestion
    [[nodiscard]] constexpr ConstReference At(SizeType aPos) const
```

wopss commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,498 @@
+        if (aPos < 0 || aPos >= Size())
+            throw std::out_of_range("DynArray::At out of range");
+
+        return Data()[static_cast<SizeType>(aPos)];
```
```suggestion
        return Data()[aPos];
```

wopss commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,498 @@
+        if (aPos < 0 || aPos >= Size())
+            throw std::out_of_range("DynArray::At out of range");
+
+        return Data()[static_cast<SizeType>(aPos)];
```
```suggestion
        return Data()[aPos];
```

wopss commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,498 @@
+
+    [[nodiscard]] constexpr Reference At(std::make_signed_t<SizeType> aPos)
+    {
+        if (aPos < 0)
+            aPos += Size();
```
Imo, we should not support this.

Mozz3d replied:
Why not?

wopss replied:
Using `std::make_signed<SizeType>` creates an unnecessary limitation here. Our `SizeType` (which is `uint32_t`) can handle arrays up to 4,294,967,295 elements, but converting it to signed (`int32_t`) cuts this in half to only 2,147,483,647, and you also have to take care of overflowing the signed type. While this might seem fine now, it's an artificial constraint which will be cause issues later (i.e. when handling arrays returned by the game). Converting between unsigned and signed can also cause problems when dealing with values above `INT32_MAX`

If we need to access elements from the end, C++ already gives us clear ways to do this with `Size() - N` or similar approaches.

Beside this issue, in my opinion, I don't see any reason to add complexity by supporting indexing patterns from other languages.

Mozz3d replied:
I think you misunderstand how exactly I implemented this, as in the event it tries to actually negatively index `Data()` it throws an out of range error. But I did overcomplicate this regardless.

I was also aware of the int32 vs uint32 max size concerns but just figured if you have an array that large, you have other concerns...

Mozz3d replied:
`rbegin` and `rend` are already there, just using standard `std::reverse_iterator` constructed from Wopss' generice iterator implementation

wopss commented on include/RED4ext/Containers/DynArray.hpp:
I  still don't get why this file is not shown as moved. It is hard to see what is added and removed in it.

Did `git mv` fail?
If `git mv` is causing issues, then it might be an idea to not move this file and let it in `include/RED4ext` directory.

wopss replied:
Ahh, I see.

wopss commented on include/RED4ext/Containers/Span.hpp:
```diff
@@ -0,0 +1,135 @@
+    {
+    }
+
+    constexpr operator bool() const noexcept
+    {
+        return IsEmpty();
+    }
```
If we are conforming to the STL, this class should not have this method.

Mozz3d replied:
You're right but Psi implemented this, so I'll coordinate with him

wopss commented on include/RED4ext/Detail/Containers/ArrayIterator.hpp:
No comment that I created this file, I'm sad :(

Mozz3d replied:
My mistake! I'll add a note

Mozz3d committed:
Several improvements & adjustments

Mozz3d committed:
Fix mistake & ArrayIterator crediting

Mozz3d committed:
Fix errors

Mozz3d committed:
Fix compiler warning

wopss commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,550 @@
+        Clear();
+        Resize(newSize);
+
+        if constexpr (std::is_trivially_copyable_v<ValueType>)
+        {
+            auto begin = *aBegin;
+            std::memcpy(Data(), &begin, newSize * sizeof(ValueType));
+        }
```
Never do a `memcpy` or `memmove`. You might have pointers in the object that can point to itself, or to another variable in the object. Doing a `copy` you will have undefined behavior.

wopss commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,550 @@
+
+    void Resize(SizeType aSize)
+    {
+        if (aSize > Capacity())
+            Reserve(aSize);
+
```
```suggestion
        if (aSize > Capacity())
        {
            Reserve(aSize);
        }
```

wopss commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,550 @@
+    }
+#pragma endregion
+
+    T* entries;        // 00
+    uint32_t capacity; // 08
+    uint32_t size;     // 0C
```
Can be private.

wopss commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,550 @@
+
+        if constexpr (std::is_trivially_copyable_v<T>)
+        {
+            std::memmove(aDst, aSrc, aCount * sizeof(ValueType));
```
Related to the previous `memmove` comment. `memcopy` and `memmove` are C constructs.

Mozz3d replied:
Are you suggesting I change this?

wopss replied:
Yes.

wopss commented on include/RED4ext/Containers/StaticArray.hpp:
```diff
@@ -105,7 +98,34 @@ struct StaticArray
     }
 #pragma endregion
-#pragma endregion
+    [[nodiscard]] constexpr Reference Front()
+    {
+        assert(!Empty());
+        return Data()[0];
+    }
+
+    [[nodiscard]] constexpr ConstReference Front() const
+    {
+        assert(!Empty());
+        return Data()[0];
+    }
+
+    [[nodiscard]] constexpr Reference Back()
+    {
+        assert(!Empty());
+        return Data()[Size() - 1];
+    }
+
+    [[nodiscard]] constexpr ConstReference Back() const
+    {
+        assert(!Empty());
+        return Data()[Size() - 1];
+    }
```
Imo, things like `begin`, `front`, etc. should only be STL style to not have duplicated. Wdyt?

Same for DynArray, Span.

Mozz3d replied:
Could you elaborate on what you mean a little? Im not sure if you're referring to just casing or something else aswell

psiberx replied:
I think it should either STL or our style, but not both, with only exceptions for compliance with named requirements, range-based for loop, etc. If we want to keep our naming, then it should be `Size()`, `Begin()`, `Front()`, etc. Things like `begin` and `end` should be in a separate "compliance" region, not mixed in with the rest.

wopss replied:
I don't think lowercase `front` has any purpose, but I made a typo, I wanted to say `being`, `end`, etc. Basically, everything  that is needed for range-based loop.

For these requirements, I would say to just use the lowercase (STL variant) naming, instead of having both.

psiberx replied:
I'd keep `Begin` and `End` for consistent naming (you still need to use them explicitly) and not consider `begin` and `end` as part of our interface.

psiberx replied:
Consistent across the class at least, ideally across the project. `begin` and `end` should not be considered functions of a class, but unfortunately C++ doesn't have a way to handle it properly. For example, C# supports explicit interface implementation that would hide `begin` and `end` from class.

Mozz3d commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -47,56 +53,85 @@ struct DynArray
+        aOther.size = 0;
+    }
+
+    template<ContainerType TContainer>
+    DynArray(const TContainer& aContainer, Memory::IAllocator* aAllocator = nullptr)
+        : DynArray(aAllocator)
+    {
+        Assign(aContainer.begin(), aContainer.end());
+    }
+
+    template<ContainerType TContainer>
+    DynArray(TContainer&& aContainer, Memory::IAllocator* aAllocator = nullptr)
+        : DynArray(aAllocator)
```
I gave some of my reasons in discord, but yes I think so. Also the template using the `ContainerType` concept to enforce what exactly can be used here, so things like `Map` aren't a concern afaict

wopss replied:
Why not do something like:
```cpp
template<typename InputIt>
requires std::input_iterator<InputIt>
DynArray(InputIt aFirst, InputIt aLast, Memory::IAllocator* aAllocator = nullptr)
```
?

Mozz3d replied:
Because I have a habit of over-complicating things when I haven't slept :)
Thank you

Mozz3d committed:
Several adjustments

Mozz3d committed:
Move SortedArray.hpp

Mozz3d committed:
Adjust include in TweakDB.hpp

Mozz3d committed:
Fixes and cleanup

Mozz3d committed:
More adjustments

psiberx commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,518 @@
+    [[nodiscard]] constexpr SizeType MaxSize() const noexcept
+    {
+#undef max // preprocessor thinks max is a macro under NOMINMAX
+        return std::numeric_limits<SizeType>::max();
```
I think the correct limit is `std::min(std::numeric_limits<SizeType>::max(), std::numeric_limits<uint32_t>::max() / sizeof(ValueType))` because allocator interface only allows us to assign up to `std::numeric_limits<uint32_t>::max()` bytes.
This can be `static constexpr`, and this method returns it.

psiberx commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,518 @@
+        return size;
+    }
+
+#pragma region STL
```
We only need `begin` and `end` for [range-based for loop](https://en.cppreference.com/w/cpp/language/range-for).

psiberx commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,518 @@
+                     void (*aMoveFunc)(Pointer aDstBuffer, Pointer aSrcBuffer, int32_t aSrcSize, DynArray* aSrcArray));
+
+        static UniversalRelocFunc<func_t> func(Detail::AddressHashes::DynArray_Realloc);
+        func(this, aNewCapacity, sizeof(ValueType), alignment >= 8 ? alignment : 8, nullptr);
```
The last argument must be a pointer to function that moves elements to a new buffer. When `nullptr` is passed, it just copies memory to a new location.

psiberx commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,518 @@
+        return *this;
+    }
+
+    constexpr Reference operator[](SizeType aPos)
```
Why this method and many others are marked as `constexpr`?

wopss replied:
That's how STL define `std::vector` (https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1004r2.pdf) . This should allow the class to be used in `constexpr` functions.

psiberx replied:
But our implementation is not `constexpr`. AFAIK even `std::vector` is still limited to transient allocations inside a constant expression, i.e. it must be destroyed before constant expression evaluation ends. We also won't be able to make our allocators `constexpr` without falling back to `std::allocator` (but it's not worth the effort, you can just use `std::vector` instead, it won't get to runtime anyway).

wopss replied:
Ah true, I forgot about the allocator being fixed. Then we can just remove it.

Mozz3d commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,518 @@
+
+        Resize(static_cast<SizeType>(diff));
+
+        if constexpr (std::is_move_constructible_v<ValueType>)
+        {
+            std::move(aFirst, aLast, Data());
+        }
+        else
+        {
+            std::copy(aFirst, aLast, Data());
+        }
```
You're right about it not making sense, not sure why I did it that way. But do we not want to support move semantics/move iterators in `Assign`?
I could just fix this to behave as inteded

psiberx replied:
There's no way to bring move semantics here and move iterators don't need special handling.

Mozz3d committed:
Several more improvments

- Fixed issues
- Added more methods (Erase, Insert, etc.)
- Refactored various methods
- Improved semantics
- Generally cleaned up

Mozz3d committed:
Some fixes and minor improvement

psiberx commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,592 @@
+    {
+        assert(Contains(aPos));
+
+        auto itRange = std::distance(aFirst, aLast);
```
`distance`

psiberx commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,592 @@
+        return Find(aValue) != End();
+    }
+
+    [[nodiscard]] bool Contains(ConstIterator aPos) const noexcept
+    {
+        return Begin() <= aPos && aPos <= End();
+    }
+
+    [[nodiscard]] bool Contains(ConstIterator aFirst, ConstIterator aLast) const noexcept
+    {
+        return Begin() <= (std::min)(aFirst, aLast) && (std::max)(aLast, aFirst) <= End();
+    }
```
I'd rename these to `Includes` or similar. Maybe make private?

Mozz3d replied:
Personally, I think `Contains` is fine and expected naming for such a method. Also, why make private? I get that they're used for checks but I don't see why we can't have them exposed for convenience

psiberx replied:
It's not expected to accept iterators. Contains is expected to work with keys or values. For iterators/ranges there's `std::includes`.

psiberx commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,592 @@
+        return Begin() + posIdx;
+    }
+
+    template<typename U>
+    Iterator Insert(ConstIterator aPos, U&& aValue)
```
Why `U` instead of `ConstReference`?

Mozz3d committed:
DynArray Buffer Move & Semantic Improvements

Mozz3d committed:
Fix errors

Mozz3d committed:
Use `MoveEntries` only if `ValueType` is not trivially relocatable

Mozz3d committed:
Update `DynArray_Realloc` signature

Mozz3d committed:
Fix instantiation (I think?)

Mozz3d committed:
Fix syntax

Mozz3d committed:
Fix my sheer stupidity

Mozz3d committed:
Rename iterator based `Contains` to `Includes` & make private

psiberx commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,584 @@
+        if (aSrcSizeInBytes == 0)
+            return;
+
+        std::move(aSrcBuffer, aSrcBuffer + aSrcArray->Size(), aDstBuffer);
+        std::destroy(aSrcBuffer, aSrcBuffer + aSrcArray->Size());
```
It can be `aSrcArray->m_size`.

wopss replied:
I would say  `Size` should be called here and the class' member fields be only used in their respective methods.

psiberx commented on include/RED4ext/Containers/StaticArray.hpp:
```diff
@@ -0,0 +1,275 @@
+        return Begin() <= (std::min)(aFirst, aLast) && (std::max)(aLast, aFirst) <= End();
+    }
+
+    void Resize(SizeType aNewSize)
+    {
+        if (aNewSize > MaxSize())
+            throw std::invalid_argument("StaticArray::Resize: New size cannot exceed MaxSize");
+
+        if (aNewSize == size)
+            return;
+
+        size = aNewSize;
+    }
```
It doesn't destroy elements. For example, if you have static array of handles (shared ptrs) and you clear the array, the handles will not be destroyed and will keep the object alive while being inaccessible.

psiberx commented on include/RED4ext/Containers/StaticArray.hpp:
```diff
@@ -0,0 +1,275 @@
+        return Find(aValue) != End();
+    }
+
+    [[nodiscard]] bool Contains(ConstIterator aPos) const noexcept
+    {
+        return Begin() <= aPos && aPos <= End();
+    }
+
+    [[nodiscard]] bool Contains(ConstIterator aFirst, ConstIterator aLast) const noexcept
+    {
+        return Begin() <= (std::min)(aFirst, aLast) && (std::max)(aLast, aFirst) <= End();
+    }
```
`Includes`

Mozz3d committed:
Fix StaticArray::Resize & improve empty check semantics

Mozz3d committed:
Merge remote-tracking branch 'upstream/master' into pr/180

Mozz3d commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,584 @@
+        , m_size(aOther.m_size)
+        , m_capacity(aOther.m_capacity)
+    {
+        aOther.m_entries = *std::bit_cast<Pointer*>(aOther.GetAllocator());
```
Are you asking what is returned by `GetAllocator`?

wopss replied:
What was the conclusion of this? Imo, the allocator should already be present when you move the `entries`.

psiberx replied:
It's correct. When DynArray is blank, `entries` stores a complete allocator instance (8 bytes), we are "resetting" `entries` for moved array here.

wopss replied:
Ah right, my brain completely ignored `aOther`.

Mozz3d commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,584 @@
+
+    DynArray& operator=(DynArray&& aOther) noexcept
+    {
+        if (this != std::addressof(aOther))
```
I believe it was @wopss who wanted it this way, as it doesn't introduce any complexity

wopss replied:
I don't remember complaining about this. I also don't get why we don't do `std::swap`.

Mozz3d commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,584 @@
+    }
+
+    template<std::input_iterator InputIt>
+    void Assign(InputIt aFirst, InputIt aLast)
```
What are you suggesting here? Should I create a new method to accept a range, or adjust this one? If it's the former, what structures are implied by ranges and views?

wopss commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,584 @@
+
+    [[nodiscard]] constexpr SizeType MaxSize() const noexcept
+    {
+#undef max // avoid conflict with minmax
```
I would be against providing `NOMINMAX` macro since others might not like it when this library is used in their code.

WSSDude replied:
Believe it is pretty universal tbh, and you already do WIN32_LEAN_AND_MEAN I believe but would have to double-check. Unless in C, those macros are pointless and just cause issues like the above to appear. Macro is kept only for backwards compar am fairly certain and safe to provide. Always can be undefed by user.

WSSDude replied:
Id be for this still.

wopss replied:
`WIN32_LEAN_AND_MEAN` is a private define, not a public one. What you're saying is to have a `NOMINMAX` as public, which is what I'm against. This should be done by the consumer, not by our library.

wopss commented on include/RED4ext/DynArray.hpp:
I just thought about this, but removing this header and class will cause a lot of issues. So, I propose to add it back, deprecate them (the header and the class).

Mozz3d replied:
What kind of issues? The changes already include updates to all the DynArray file paths, and `Reflection-inl.hpp` was updated accordingly for the dumper

wopss replied:
People project's still use this path, changing it like that will be a breaking change.

Mozz3d commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,584 @@
+        }
+
+        auto distance = std::distance(aFirst, aLast);
+        DifferenceType newSize = std::abs(distance);
```
How do you suggest I avoid the double evaluation?
I assume it involves changing the template concept to `std::forward_iterator`?

Mozz3d replied:
Would `std::back_inserter` work like you suggest? DynArray doesn't currently (STL semantically) define `push_back`

WSSDude replied:
It wouldnt that is true, there should be push_back also in the STL portion of the containers, along with value_type typedef so you can use std::ranges::value_type

Mozz3d committed:
Merge branch 'master' into master

Mozz3d committed:
Suggested changes

- Added range based container assignment and construction

- Replaced `=` operator to use `std::swap` of members

- Improved semantics

- Deprecated `RED4ext/DynArray.hpp` with pragma message warning

Mozz3d committed:
Merge branch 'master' into master

WSSDude commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,584 @@
+
+    void PushBack(ValueType&& aItem)
+    {
+        EmplaceBack(std::forward<ValueType&&>(aItem));
```
They are not the same, && also eats const &, when you move instead of forward, that info is lost and leads to undefined behaviour. In this instance, it should be forward as you forward the argument further down the pipeline.

Mozz3d committed:
Semantic improvments

- STL compatibility
- Swap method
- Removed negative range handling

Mozz3d committed:
Merge branch 'master' of https://github.com/Mozz3d/RED4ext.SDK

Mozz3d committed:
STL compatibility for `Span` & `StaticArray`

Mozz3d committed:
Prefix member names in `StaticArray`

wopss commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,621 @@
+    }
+#pragma endregion
+private:
+    T* m_entries;        // 00
+    uint32_t m_capacity; // 08
+    uint32_t m_size;     // 0C
```
Move these after the functions.

wopss commented on include/RED4ext/Containers/Span.hpp:
```diff
@@ -0,0 +1,232 @@
+        return Find(aValue) != End();
+    }
+
+    [[nodiscard]] bool Contains(ConstIterator aPos) const noexcept
+    {
+        return Begin() <= aPos && aPos <= End();
+    }
+
+    [[nodiscard]] bool Contains(ConstIterator aFirst, ConstIterator aLast) const noexcept
+    {
+        return Begin() <= (std::min)(aFirst, aLast) && (std::max)(aLast, aFirst) <= End();
+    }
```
When I think about `Contains`, I  would assume it is that the Span contains the value, not the iterator. I would say this is `IsInRange`. 

Besides this, is this function really necessary?

wopss commented on include/RED4ext/Containers/StaticArray.hpp:
```diff
@@ -0,0 +1,304 @@
+
+namespace RED4ext
+{
+template<typename T, uint32_t MAX_LEN>
```
```suggestion
template<typename T, uint32_t Length>
```

Mozz3d replied:
I think `MAX_LEN` makes sense here and should stay, or at least be some other name that indicates that this is the max amount of space being given to the array

wopss replied:
That's also good.

wopss replied:
That's also good.

wopss commented on include/RED4ext/Containers/StaticArray.hpp:
```diff
@@ -0,0 +1,304 @@
+namespace RED4ext
+{
+template<typename T, uint32_t MAX_LEN>
+struct StaticArray
```
Looking at the code, this is not a static array, but a dynamic array with stack allocated entries.

wopss replied:
StaticArray is a fixed size array, you cannot change its size after creation. That's what most languages/frameworks/engines do.

I don't remember any specific functions in the game's RTTI about StaticArrays. How does redscript handle them?

wopss replied:
My point is that `m_size` is constant throughout the life of the array, so methods like: `Resize` and `Clear`, are not needed.

Mozz3d replied:
That's not the case here, `StaticArray`s do have a dynamic size, constrained to be within their defined max size. This is evident throughout the engine. I think you're confusing it with `FixedArray`, which also exists

wopss replied:
@psiberx can you confirm?

psiberx replied:
Yes, this is how it works.

wopss commented on include/RED4ext/DynArray.hpp:
```diff
@@ -1,357 +1,4 @@
 #pragma once
+#pragma message("WARNING: RED4ext/DynArray.hpp is deprecated. Please use RED4ext/Containers/DynArray.hpp instead.")
```
```suggestion
#pragma message("WARNING: 'RED4ext/DynArray.hpp' is deprecated. Please use 'RED4ext/Containers/DynArray.hpp' instead.")
```

psiberx commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,621 @@
+        if (m_capacity)
+        {
+            Clear();
+            auto allocator = *std::bit_cast<Pointer*>(GetAllocator());
+            std::bit_cast<Memory::IAllocator*>(&allocator)->Free(m_entries);
+            m_entries = allocator;
```
It's technically correct, but they do this in the engine. I couldn't recall the exact case unfortunately, but there was an issue related to this. I'd keep it with a comment that it matches the engine behavior.

Mozz3d commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,621 @@
+    template<class... TArgs>
+    Iterator Emplace(ConstIterator aPos, TArgs&&... aArgs)
+    {
+        assert(Includes(aPos));
```
It was previously `Contains` but @psiberx suggested the rename to `Includes` [here](https://github.com/wopss/RED4ext.SDK/pull/180#discussion_r2044582903) 
They were intended for convenience, but at this rate, I can remove them altogether and let the asserts just use `std::find` directly, would that be ideal?

psiberx replied:
Let's use `IsInRange`.

Mozz3d commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,621 @@
+        SizeType newSize = m_size + 1;
+        Reserve(newSize);
+
+        SizeType tailSize = m_size - posIdx;
+        if (tailSize > 0)
+            ShiftEntries(Iterator(&m_entries[posIdx]), Iterator(&m_entries[posIdx + 1]), tailSize);
+
+        new (&m_entries[posIdx]) ValueType(std::forward<TArgs>(aArgs)...);
+
+        m_size = newSize;
+        return Begin() + posIdx;
```
The allocator is moved as a result of `SetCapacity`, yes. It becomes `Begin() + m_capacity`

Mozz3d commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,621 @@
+        else
+        {
+            Reserve(aNewSize);
+            std::uninitialized_default_construct(End(), Begin() + aNewSize);
```
Yes, the native call modifies capacity, not size.

wopss commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,621 @@
+        SizeType newSize = m_size + 1;
+        Reserve(newSize);
+
+        SizeType tailSize = m_size - posIdx;
```
That was more like a nitpick btw, feel free to ignore it.

wopss commented on include/RED4ext/Containers/StaticArray.hpp:
```diff
@@ -0,0 +1,304 @@
+    T m_entries[MAX_LEN]; // 00
+    uint32_t m_size;
+
+    [[nodiscard]] bool Includes(ConstIterator aPos) const noexcept
+    {
+        return Begin() <= aPos && aPos <= End();
+    }
+
+    [[nodiscard]] bool Includes(ConstIterator aFirst, ConstIterator aLast) const noexcept
+    {
+        return Begin() <= (std::min)(aFirst, aLast) && (std::max)(aLast, aFirst) <= End();
+    }
```
`IsInRange` name change here too.

Mozz3d committed:
Suggested changes

Mozz3d committed:
Remove duplicate `Find` definitions

Mozz3d committed:
Keep `min` `max` macros undefine

Mozz3d committed:
Keep const and non-const `Find` implementations

Mozz3d committed:
Keep explicit `ConstIterator` cast in `DynArray::Emplace`

Mozz3d committed:
Fix naming in `StaticArray`

With additional fix for `TagList::Contains`

Mozz3d committed:
Fix duplicate `Find` definition in `StaticArray`

with another fix for `TagList` to use the correct method

Mozz3d committed:
Remove whitespace

Mozz3d committed:
Try wrapping `std::min` to avoid macro conflict

Mozz3d committed:
Try again

Mozz3d committed:
Appease clang formatting

Mozz3d committed:
Update dumps for 2.3

Mozz3d committed:
Remove accidental CET dump

Mozz3d commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,618 @@
+#pragma once
+
+#include <algorithm>
```
I think Clang enforces these up here, regardless this style is present in all other files with similar includes

Mozz3d committed:
Move generic includes down

Mozz3d committed:
Merge branch 'master' into master

Mozz3d committed:
Fix use of outdated method in `TagList::Contains`

psiberx commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,595 @@
+#undef max // avoid conflict with minmax
+#undef min
+        static constexpr auto maxSize =
+            std::min(std::numeric_limits<SizeType>::max(),
```
This should be `(std::min)` like in other places.

wopss commented on include/RED4ext/Containers/StaticArray.hpp:
```diff
@@ -110,27 +110,27 @@ struct StaticArray
     }
 
-    [[nodiscard]] bool Contains(ConstReference aValue) const noexcept
+    [[nodiscard]] bool IsInRange(ConstReference aValue) const noexcept
```
Remove `noexcept`, `Find` is not marked as `noexcept`.

Mozz3d commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -506,35 +509,6 @@ struct DynArray
     }
 
 #pragma region STL
-    using value_type = ValueType;
```
I wish I could ask my past self the same question

Mozz3d committed:
Improve semantics

Mozz3d committed:
Avoid `max` macro conflict

Mozz3d committed:
Use explicit `SizeType` casts & improve constness

Mozz3d committed:
Outdated method name

Mozz3d committed:
Too lazy to reinstall clang prehook so this is what I get

wopss reviewed (approved):
The only item left from me is: https://github.com/wopss/RED4ext.SDK/pull/180#discussion_r2686411169.

But I do not care that much for now.

psiberx commented on include/RED4ext/Containers/StaticArray.hpp:
```diff
@@ -110,27 +110,27 @@ struct StaticArray
     void Resize(SizeType aNewSize)
     {
-        if (aNewSize == size)
+        if (aNewSize == m_size)
```
The sole purpose of this getter is to provide public access to the data, not to access the data in general; this data doesn't need access logic. Using a public getter in an inner scope is misuse, implying the need for access logic.
If we want tracing and think about later changes to access logic, why there's no setter for every property as well? It's because this getter wasn't added to wrap the access to the data, it was added for a different reason -- to give public access.
IMO modern IDEs make it so easy to add a getter whenever it's really needed, that it's better to keep the current code simpler and cleaner.

Mozz3d committed:
Merge branch 'master' into master

Mozz3d committed:
Method renaming

Mozz3d committed:
Merge branch 'master' of https://github.com/Mozz3d/RED4ext.SDK

Mozz3d commented on include/RED4ext/Containers/DynArray.hpp:
```diff
@@ -0,0 +1,612 @@
+struct IAllocator;
+}
+template<typename T>
+struct DynArray
```
No, I think I made a mistake when renaming previously

Mozz3d committed:
Fix uses of `Empty`->`IsEmpty`

Mozz3d committed:
Merge remote-tracking branch 'upstream/master'

Mozz3d committed:
Merge branch 'master' into master

Mozz3d committed:
Update Changelog

Mozz3d committed:
Improve changelog according to feedback

Mozz3d committed:
Use imperative perfect tense and reduce whitespace in changelog

wopss commented on CHANGELOG.md:
```diff
@@ -8,6 +8,25 @@ and this project adheres to
 
+### Added
+
+- Added `include/RED4ext/Detail/Containers/ArrayIterator.hpp`, an iterator type
```
I don't see as a feature worth mentioning.

wopss commented on CHANGELOG.md:
```diff
@@ -8,6 +8,25 @@ and this project adheres to
+  used by `DynArray`, `StaticArray`, and `Span`.
+- Added `DynArray` constructors for input iterators and containers satisfying `std::ranges::range` concept.
+- Added several methods for manipulating `DynArray` buffer data.
+- Added `StaticArray` constructors and methods.
```
If you can explain this more, it would be great.

Mozz3d committed:
Changelog: Remove `ArrayIterator` and elaborate

Mozz3d committed:
More changelog brevity
